src/sentry/taskworker/consumers/factory.py

```diff
@@ -0,0 +1,140 @@
+            ((_partition, offset),) = message.committable.items()
+            deadletter_at = int(time.time()) + self.max_pending_timeout
+
+            return InflightActivation(
+                activation=activation,
+                status=TASK_ACTIVATION_STATUS_PENDING,
+                offset=offset,
+                added_at=Timestamp(seconds=int(time.time())),
+                deadletter_at=Timestamp(seconds=deadletter_at),
+                processing_deadline=None,
+            )
+
+        def accumulator(
+            batched_results: MutableSequence[InflightActivation],
+            message: BaseValue[InflightActivation],
+        ) -> MutableSequence[InflightActivation]:
+            batched_results.append(message.payload)
+            return batched_results
+
+        def flush_batch(
+            message: Message[MutableSequence[InflightActivation]],
+        ) -> Message[MutableSequence[InflightActivation]]:
+            logger.info("Flushing batch. Messages: %r...", len(message.payload))
+            self.inflight_task_store.store(message.value.payload)
+            return message
+
+        def do_upkeep(
+            message: Message[KafkaPayload],
+        ) -> KafkaPayload:
+            self.inflight_task_store.handle_processing_deadlines()
+            self.inflight_task_store.handle_retry_state_tasks()
+            self.inflight_task_store.handle_deadletter_at()
+            self.inflight_task_store.handle_failed_tasks()
+            self.inflight_task_store.remove_completed()
+
+            return message.payload
+
+        def limit_tasks(
+            message: Message[KafkaPayload],
+        ) -> KafkaPayload:
+            count = self.inflight_task_store.count_pending_task()
+            if count >= self.max_inflight_activation_in_store:
+                # The number of pending inflight activations in the store exceeds the limit.
+                # Wait for workers to complete tasks before adding the next offset to the queue.
+                logger.info(
+                    "Number of inflight activation: %s exceeds the limit: %s. Retrying in 3 seconds",
+                    count,
+                    self.max_inflight_activation_in_store,
+                )
+                raise MessageRejected()
+            return message.payload
+
+        flush = RunTask(
+            function=flush_batch,
+            next_step=CommitOffsets(commit),
+        )
+
+        collect = Reduce(
+            max_batch_size=self.max_batch_size,
+            max_batch_time=self.max_batch_time,
+            accumulator=accumulator,
+            initial_value=list,
+            next_step=flush,
+        )
+
+        process = RunTask(
+            function=process_message,
+            next_step=collect,
+        )
+
+        limit = RunTask(
+            function=limit_tasks,
+            next_step=process,
+        )
+
+        return RunTask(
+            function=do_upkeep,
```

Comment:
This has the upkeep happening before the message is processed. Wouldn't we want to process the message, commit the offsets, and then do upkeep?